text/0000-build-std/4-proposal.md

```diff
@@ -0,0 +1,2138 @@
+behaviour can be forced by `-Clink-self-contained=true`, but is force-enabled
+for some targets and inferred for others.
+
+Rust can start to ship `rust-self-contained` components for any targets which
+need it. These components will contain the special object files normally
+included in `rust-std`, and will be distributed for all tiers of targets. While
+generally these objects are specific to the architecture and C runtime (CRT)
+(and so `rust-self-contained-$arch-$crt` could be sufficient and result in fewer
+overall components), it's technically possible that Rust could support two
+targets with the same architecture and same CRT but different versions of the
+CRT, so having target-specific components is most future-proof. These would
+replace the `self-contained` directory in existing `rust-std` components.
+
+As long as these components have been downloaded, as well as any other support
+components, such as `rust-mingw`, rustc's `-Clink-self-contained` will be able
+to link against the object files and build-std should never fail on account of
+missing special object files.
+
+*See the following sections for future possibilities:*
+
+- [*Enable local recompilation of special object files/sanitizer runtimes*][future-recompile-special]
+
+### Compiler builtins
+[compiler-builtins]: #compiler-builtins
+
+#### `compiler-builtins-mem`
+[compiler-builtins-mem]: #compiler-builtins-mem
+
+The `mem` feature of `compiler_builtins` (and the subsequent
+`compiler-builtins-mem` feature of `core`, `alloc`, `std` which forward to
+`compiler_builtins/mem`) is required by `no_std` crates because `libc` does not
+provide these symbols without `std`.
+
+It is necessary that the `compiler-builtins-mem` feature of `alloc` and/or
+`core` be enabled when `std` is not in the crate graph
+([?][rationale-no-weak-linkage]).
+
+*See the following sections for rationale/alternatives:*
+
+- [*Why not use weak linkage for `compiler-builtins/mem` symbols?*][rationale-no-weak-linkage]
+
+#### `compiler-builtins-c`
+[compiler-builtins-c]: #compiler-builtins-c
+
+The [`c` feature][background-dependencies] of `compiler_builtins` (which is also
+exposed by `core`, `alloc` and `std` through `compiler-builtins-c`) causes its
+`build.rs` file to build and link in more optimised C versions of the intrinsics
+it implements.
+
+It will not be enabled by default because it is possible that the target
+platform does not have a suitable C compiler available. The user being
+able to enable this manually will be enabled through work on
+features (see
+[*Allow enabling/disabling features with build-std*][future-features]).
+
+The default behaviour of `compiler_builtins` can and should be improved to allow
+`build-std` to match the prebuilt standard library as much as possible. The
+`compiler-rt` sources should be distributed with `rust-src` (as `libunwind` is
+today). By default `compiler_builtin`'s `build.rs` should make a best-effort
```

Comment:
I think this section should be removed and we have to accept that by default build-std will often perform worse that prebuilt for now.

The issue is that when a target-modifier is changed, `CFLAGS` (which should be advertised as being respected) needs to be changed at the same time. If not then the best case scenario for the user is a big meaty linker error and the worst case is some kind of CFI not being fully applied. I'm not sure how otherwise to make the best-effort approach work.